### PR TITLE
结合 NWR 实现 Flexible raft

### PR DESCRIPTION
**Pull request by 1294566108**
### Motivation:
ospp开源之夏-2023-结合 NWR 实现 Flexible raft，用于自定义 Quorum 的大小

### Modification:

新增QuorumNWQ设计文档至rfcs

### Result:

Fixes #<1003>.

### CODE REVIEW

**Comment by shihuili1218** [comment]
辛苦啦，rfc可以单独提交pr

File: rfcs/0002-Flexible-Raft-with-NWR.md [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+     * Write Quorum's factor
+     */
+    private double                          writeQuorumFactor;
+    /**
+     * Enable NWRMode or Not
+     */
+    private boolean                         enableNWRMode          = false;
+```
+
+对于readQuorumFactor和writeQuorumFactor两个属性，在NodeOptions类里提供了setter和getter方法便于用户自定义配置。对于enableNWRMode属性，提供了isEnableNWRModeI()来判断是否开启NWR模型，而enableNWRMode()方法表示开启NWR模式。
```
Comment (killme2008) [inline rfcs/0002-Flexible-Raft-with-NWR.md:64]:
```suggestion
对于readQuorumFactor和writeQuorumFactor两个属性，在NodeOptions类里提供了setter和getter方法便于用户自定义配置。对于enableNWRMode属性，提供了isEnableNWRMode()来判断是否开启NWR模型，而enableNWRMode()方法表示开启NWR模式。
```

File: rfcs/0002-Flexible-Raft-with-NWR.md [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+     */
+    private double                          readQuorumFactor;
+    /**
+     * Write Quorum's factor
+     */
+    private double                          writeQuorumFactor;
+    /**
+     * Enable NWRMode or Not
+     */
+    private boolean                         enableNWRMode          = false;
```
Comment (killme2008) [inline rfcs/0002-Flexible-Raft-with-NWR.md:61]: Maybe `enableFlexibleRaft` is more proper.

File: rfcs/0002-Flexible-Raft-with-NWR.md [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+     return false;
+}
+```
+
+校验与同步逻辑在方法checkAndResetFactor里：
+
+```
+    private boolean checkAndResetFactor(Double writeFactor, Double readFactor){
+        if (Objects.nonNull(readFactor) && Objects.nonNull(writeFactor)) {
+            if (readFactor + writeFactor != 1) {
```
Comment (killme2008) [inline rfcs/0002-Flexible-Raft-with-NWR.md:106]: We can't compare integer type(`1`) with float type(`readFactor + writeFactor`) precisely. I think we can change the `readFactor` and `writeFactor` to be int types, and ensure that their sum is `10`. Their range both are `(0,10)`.

File: rfcs/0002-Flexible-Raft-with-NWR.md [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+
+移除成员的逻辑是：基于原conf生成一份新conf，这份新配置移除了指定的peer，然后调用unsafeRegisterConfChange方法执行成员配置变更，对于移除成员，直接调用nextStage()进入下一阶段记录成员变更日志项。
+
+##### changePeers
+
+与前两者方法类似。不过对于处理多成员变更，也是需要进行逐个成员变更的，也就是所谓的单节点变更。
+
+##### resetPeers
+
+这个方法用于强制变更本节点的配置，单独重置该节点的配置，而在该节点成为领导者之前，无需复制其他同行。 当复制组的大多数已死时，应该调用此功能。在这种情况下，一致性和共识都不能保证，在处理此方法时要小心。
```
Comment (killme2008) [inline rfcs/0002-Flexible-Raft-with-NWR.md:490]: stepDown 的逻辑不需要改吗？ 毕竟多数派的定义变了。

File: rfcs/0002-Flexible-Raft-with-NWR.md (hunk 3 above)
Comment (1294566108) [reply rfcs/0002-Flexible-Raft-with-NWR.md:106]: Does this mean that if the user sets **readFactor to 4** and **writeFactor to 6**. If in a cluster with **5 nodes**, _writeQuorum_ as **5 * 0.6=3** and _readQuorum_ as **n-w+1=5-3+1=3**

File: rfcs/0002-Flexible-Raft-with-NWR.md (hunk 4 above)
Comment (shihuili1218) [reply rfcs/0002-Flexible-Raft-with-NWR.md:490]: > 

好像没有发现这里需要修改的地方吧，你是指nwr的数据存储吗？类似voteFor一样！

File: rfcs/0002-Flexible-Raft-with-NWR.md (hunk 3 above)
Comment (killme2008) [reply rfcs/0002-Flexible-Raft-with-NWR.md:106]: Yes， that's my point @1294566108

File: rfcs/0002-Flexible-Raft-with-NWR.md (hunk 4 above)
Comment (killme2008) [reply rfcs/0002-Flexible-Raft-with-NWR.md]: 我看他已经改了，是对多数派存活的判断逻辑需要改下。 @shihuili1218

**Pull request closed without merging**